astropy/io/fits/tests/test_hdulist.py

```diff
@@ -550,6 +550,23 @@ def test_replace_memmaped_array(self):
         with fits.open(self.temp('temp.fits'), memmap=True) as hdul:
             assert ((old_data + 1) == hdul[1].data).all()
 
+    @pytest.mark.remote_data
+    def test_open_file_with_bad_file_padding(self):
+        """
+        Test warning when opening files with extra padding at the end.
+        See https://github.com/astropy/astropy/issues/4351
+        """
+
+        with pytest.warns(
+            AstropyUserWarning,
+            match='Unexpected extra padding at the end of the file.'
+        ) as w:
+            fits.open(
+                'https://dataverse.harvard.edu/api/access/datafile/2439198',
+                cache=False
+            )
```

Comment:
The reason this test works for you interactively but not in the test suite, as that `fits.open()` doesn't actually immediately try to read the whole file (it is lazy).

When you do this at the interactive prompt it tries to represent the resulting HDUList object; i.e. it is the calling the equivalent of

```
_ = fits.open('...')
repr(_)
```

whereas here the return value of `fits.open` is simply ignored and not used.  Try instead calling something that will force it to read the full file, like `fits.info()`.  See also my earlier comment on how you can rewrite this test to reproduce the problem with an arbitrary file.

Reply:
**Update:** It turns out, ironically, calling `fits.info()` when the file contains trailing bytes does *not* reproduce the error, and instead results in the warning we want.  I have to find out what is happening differently when calling `fits.info()`.

Reply:
It turns out `fits.info()` passes `ignore_missing_end=True` by default, so it doesn't reproduce the error.  Instead, try something like `len(fits.open(self.temp('temp.fits')))`.  That does the trick.

Reply:
Thank you very much for all your explanations. I have modified the test accordingly.

Reply:
I have modified a bit your suggestion because in my previous version (without the with block), a `RessourceWarning` was thrown (the file wasn't closed).

Reply:
You are right, of course.